amm/drink-tests/src/stable_swap_tests/tests_swap_received.rs

```diff
@@ -0,0 +1,255 @@
+        bob(),
+    )
+    .result
+    .unwrap()
+    .unwrap();
+    assert_eq!(
+        total_lp_required - lp_fee_part,
+        protocol_fee_lp,
+        "Incorrect protocol fee"
+    );
+}
+
+// ref https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L744
+#[drink::test]
+fn test_01(mut session: Session) {
+    setup_test_swap_received(
+        &mut session,
+        vec![6, 6],                       // decimals
+        vec![100000000000, 100000000000], // initial reserves
+        1000,                         // A
+        6,                            // fee BPS
+        2000,                         // protocol fee BPS
+        10000000000,                  // swap_amount_in
+        Ok(9999495232),               // expected out (with fee)
+    );
+}
+
+// ref https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L763
+#[drink::test]
+fn test_02(mut session: Session) {
+    setup_test_swap_received(
+        &mut session,
+        vec![12, 18],
+        vec![100000000000000000, 100000000000000000000000],
+        1000,
+        6,
+        2000,
+        10000000000000000,
+        Ok(9999495232752197989995),
+    );
+}
+
+// ref https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L782
+#[drink::test]
+fn test_03(mut session: Session) {
+    setup_test_swap_received(
+        &mut session,
+        vec![6, 6],
+        vec![100000000000, 100000000000],
+        1000,
+        6,
+        2000,
+        0,
+        Err(StablePoolError::InsufficientInputAmount()),
+    );
+}
+
+// ref https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L801
+#[drink::test]
+fn test_04(mut session: Session) {
+    setup_test_swap_received(
+        &mut session,
+        vec![12, 18],
+        vec![100000000000000000, 100000000000000000000000],
+        1000,
+        6,
+        2000,
+        0,
+        Err(StablePoolError::InsufficientInputAmount()),
+    );
+}
+
+// ref https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L820
+#[drink::test]
+fn test_05(mut session: Session) {
+    setup_test_swap_received(
+        &mut session,
+        vec![6, 6],
+        vec![100000000000, 100000000000],
+        1000,
+        6,
+        2000,
+        1,
+        Ok(0),
+    );
+}
+
+// ref https://github.com/ref-finance/ref-contracts/blob/d241d7aeaa6250937b160d56e5c4b5b48d9d97f7/ref-exchange/src/stable_swap/mod.rs#L839
+#[drink::test]
+fn test_06(mut session: Session) {
+    setup_test_swap_received(
+        &mut session,
+        vec![18, 12],
+        vec![100000000000000000000000, 100000000000000000],
+        1000,
+        6,
+        2000,
+        1000000,
+        Ok(0),
```

Comment:
In Ref's test it tries to trade 1 token, here we trade `1000000` – is the output still 0?

Reply:
It's because of the difference in token decimals. Our contract supports tokens with up to 18 decimals, whereas Ref supports up to 24 decimals with 18 decimal internal precision thus sacrificing tokens precisions. So actually our test and the referenced test check slightly different things. I should have made a note about that.
And yes, the output should be 0 since here it shows how it would work if we tried to swap `0.000000000001` DAI(18 dec) for some 12 dec token (which smallest indivisible unit would be `0.000000000001`)

Reply:
I see, thanks for the explanation.